### PR TITLE
matchbox: raw ignition support

### PR DESCRIPTION
**Pull request by mcuadros**
In order to support raw ignition files, this PR introduce a new option `raw_ignition` to the `profile_resource`.

This allows to use the builtin ignition provider:

```hcl
data "ignition_systemd_unit" "example" {
  name = "example.service"
  content = "[Service]\nType=oneshot\nExecStart=/usr/bin/echo Hello World\n\n[Install]\nWantedBy=multi-user.target"
}

data "ignition_config" "example" {
  systemd = [
    "${data.ignition_systemd_unit.example.id}",
  ]
}

resource "profile_resource" "example" {
  # ...

  raw_ignition = "${data.ignition_config.example.rendered}"
}
```

If the #11 is merged I will be happy to add test for this functionality

### CODE REVIEW

**Comment by mcuadros** [comment]
Fixes https://github.com/coreos/bugs/issues/1922
/cc @dghubble

**Comment by dghubble** [comment]
Supporting Container Linux Configs only was discussed in design. Users should provide Container Linux Configs which are explicitly designed to be human readable, human writable, and to be processed through tools which do validations for humans (return line numbers where errors occur, etc). Ignition configs are the machine representation and are more of an internal detail.

**Comment by mcuadros** [comment]
As you can see in the example the ignition is not generated by hand, is generated by the builtin provider. Also the validation is at the terraform planning phase so you don't need to worry about the server side validation. 

> If you prefer to design your own templating solution, raw Ignition files (suffixed with .ign or .ignition) are served directly. 
From https://github.com/coreos/matchbox/blob/master/Documentation/container-linux-config.md

**Comment by dghubble** [comment]
Discussing on the issue https://github.com/coreos/bugs/issues/1922#issuecomment-303205494

File: matchbox/resource_profile.go [hunk 1]
```diff
@@ -156,6 +163,19 @@ func resourceProfileDelete(d *schema.ResourceData, meta interface{}) error {
...
 }
 
-func containerLinuxConfigName(name string) string {
-	return name + ".yaml.tmpl"
+func containerLinuxConfigName(d *schema.ResourceData) string {
+	filename, _ := containerLinuxConfig(d)
+	return filename
+}
+
+func containerLinuxConfig(d *schema.ResourceData) (filename, config string) {
```
Comment (dghubble) [inline matchbox/resource_profile.go]: Should error if neither format is provided. Also, you can probably simplify this,

```
name := d.Get("name").(string)
if content, ok := d.GetOk("raw_ignition"); ok {
    return fmt.Sprintf("%s.ign", name), content.(string)
}
...
```

File: matchbox/resource_profile.go [hunk 2]
```diff
@@ -57,9 +63,13 @@ func resourceProfileCreate(d *schema.ResourceData, meta interface{}) error {
 	client := meta.(*matchbox.Client)
 	ctx := context.TODO()
 
+	if d.HasChange("raw_ignition") && d.HasChange("container_linux_config") {
+		return errors.New("raw_ignition and container_linux_config are mutual exclusive")
```
Comment (dghubble) [inline matchbox/resource_profile.go]: List "container_linux_config (recommended)" first. s/mutal/mutually.

File: matchbox/resource_profile.go [hunk 3]
```diff
@@ -57,9 +63,13 @@ func resourceProfileCreate(d *schema.ResourceData, meta interface{}) error {
 	client := meta.(*matchbox.Client)
 	ctx := context.TODO()
 
+	if d.HasChange("raw_ignition") && d.HasChange("container_linux_config") {
```
Comment (dghubble) [inline matchbox/resource_profile.go]: Hm, is it possible to get in a state where both are defined but both don't have changes? We should error if both are non-empty.

File: matchbox/resource_profile.go (hunk 1 above)
Comment (mcuadros) [inline matchbox/resource_profile.go]: Currently `container_linux_config` is Optional, but looks that make sense make it required.

As you can see I updated the PR adding a `validateResourceProfile` that validated that if both options are used or none triggering an error.

**Comment by mcuadros** [comment]
@dghubble I just added the test.

BTW Is very annoying get all this log lines from the stopping gRPC, I will try to take a look for at least hide the output.

**Comment by dghubble** [comment]
LGTM with a rebase and squash to sort out that merge commit.

**Comment by dghubble** [comment]
Rebased and squashed as 011ced7dc8e33e8b45e73c8e1c560bebaa2dfade

**Pull request closed without merging**